File: app/src/main/java/com/example/myaac/data/repository/SettingsRepository.kt
package com.example.myaac.data.repository

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import com.example.myaac.model.AppShortcut
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

data class AppSettings(
    val languageCode: String = "en",
    val textScale: Float = 1.0f,
    val ttsRate: Float = 1.0f,
    val disabilityType: com.example.myaac.data.model.DisabilityType = com.example.myaac.data.model.DisabilityType.NONE,
    val locationSuggestionsEnabled: Boolean = true,
    val showHorizontalNavigation: Boolean = false,
    val showSymbolsInSentenceBar: Boolean = false,
    val itemsToGenerate: Int = 20,
    val maxBoardCapacity: Int = 500,
    val fontFamily: String = "System",
    val displayScale: Float = 1.0f,
    val useSystemSettings: Boolean = true,
    val symbolLibrary: String = "ARASAAC",
    val showHomeOnStartup: Boolean = true,
    val homeButtonsJson: String = "",
    val appShortcutsJson: String = "",
    // Prediction settings
    val predictionEnabled: Boolean = true,
    val aiPredictionEnabled: Boolean = true,
    val predictionCount: Int = 5,
    val showSymbolsInPredictions: Boolean = true,
    val learnFromUsage: Boolean = true,
    // Grammar settings
    val autoGrammarCheck: Boolean = false
)

class SettingsRepository(context: Context) {
    private val preferences: SharedPreferences = 
        context.getSharedPreferences("app_settings", Context.MODE_PRIVATE)
    
    private val gson = Gson()
    
    private val _settings = MutableStateFlow(loadSettings())
    val settings: StateFlow<AppSettings> = _settings.asStateFlow()
    
    private fun loadSettings(): AppSettings {
        val disabilityTypeString = preferences.getString(KEY_DISABILITY_TYPE, "NONE") ?: "NONE"
        val disabilityType = try {
            com.example.myaac.data.model.DisabilityType.valueOf(disabilityTypeString)
        } catch (e: IllegalArgumentException) {
            com.example.myaac.data.model.DisabilityType.NONE
        }

        // Backward compatibility: if old maxBoardItems exists, use it for both new settings
        val legacyMaxItems = preferences.getInt("max_board_items", -1)
        val defaultGenerate = if (legacyMaxItems != -1) legacyMaxItems else 20
        val defaultCapacity = if (legacyMaxItems != -1) (legacyMaxItems * 5).coerceAtMost(500) else 500

        return AppSettings(
            languageCode = preferences.getString(KEY_LANGUAGE, "en") ?: "en",
            textScale = preferences.getFloat(KEY_TEXT_SCALE, 1.0f),
            ttsRate = preferences.getFloat(KEY_TTS_RATE, 1.0f),
            disabilityType = disabilityType,
            locationSuggestionsEnabled = preferences.getBoolean(KEY_LOCATION_SUGGESTIONS, true),
            showHorizontalNavigation = preferences.getBoolean(KEY_HORIZONTAL_NAVIGATION, false),
            showSymbolsInSentenceBar = preferences.getBoolean(KEY_SHOW_SYMBOLS_IN_SENTENCE, false),
            itemsToGenerate = preferences.getInt(KEY_ITEMS_TO_GENERATE, defaultGenerate).coerceIn(1, 50),
            maxBoardCapacity = preferences.getInt(KEY_MAX_BOARD_CAPACITY, defaultCapacity).coerceIn(50, 1000),
            fontFamily = preferences.getString(KEY_FONT_FAMILY, "System") ?: "System",
            displayScale = preferences.getFloat(KEY_DISPLAY_SCALE, 1.0f),
            useSystemSettings = preferences.getBoolean(KEY_USE_SYSTEM_SETTINGS, true),
            symbolLibrary = preferences.getString(KEY_SYMBOL_LIBRARY, "ARASAAC") ?: "ARASAAC",
            showHomeOnStartup = preferences.getBoolean(KEY_SHOW_HOME_ON_STARTUP, true),
            homeButtonsJson = preferences.getString(KEY_HOME_BUTTONS, "") ?: "",
            appShortcutsJson = preferences.getString(KEY_APP_SHORTCUTS, "") ?: "",
            // Prediction settings
            predictionEnabled = preferences.getBoolean(KEY_PREDICTION_ENABLED, true),
            aiPredictionEnabled = preferences.getBoolean(KEY_AI_PREDICTION_ENABLED, true),
            predictionCount = preferences.getInt(KEY_PREDICTION_COUNT, 5).coerceIn(3, 8),
            showSymbolsInPredictions = preferences.getBoolean(KEY_SHOW_SYMBOLS_IN_PREDICTIONS, true),
            learnFromUsage = preferences.getBoolean(KEY_LEARN_FROM_USAGE, true),
            autoGrammarCheck = preferences.getBoolean(KEY_AUTO_GRAMMAR_CHECK, false)
        )
    }
    
    fun setLanguage(code: String) {
        preferences.edit().putString(KEY_LANGUAGE, code).apply()
        _settings.value = _settings.value.copy(languageCode = code)
    }
    
    fun setTextScale(scale: Float) {
        preferences.edit().putFloat(KEY_TEXT_SCALE, scale).apply()
        _settings.value = _settings.value.copy(textScale = scale)
    }

    fun setFontFamily(family: String) {
        preferences.edit().putString(KEY_FONT_FAMILY, family).apply()
        _settings.value = _settings.value.copy(fontFamily = family)
    }

    fun setDisplayScale(scale: Float) {
        preferences.edit().putFloat(KEY_DISPLAY_SCALE, scale).apply()
        _settings.value = _settings.value.copy(displayScale = scale)
    }

    fun setUseSystemSettings(useSystem: Boolean) {
        preferences.edit().putBoolean(KEY_USE_SYSTEM_SETTINGS, useSystem).apply()
        _settings.value = _settings.value.copy(useSystemSettings = useSystem)
    }
    
    fun setTtsRate(rate: Float) {
        preferences.edit().putFloat(KEY_TTS_RATE, rate).apply()
        _settings.value = _settings.value.copy(ttsRate = rate)
    }

    fun setLocationSuggestionsEnabled(enabled: Boolean) {
        preferences.edit().putBoolean(KEY_LOCATION_SUGGESTIONS, enabled).apply()
        _settings.value = _settings.value.copy(locationSuggestionsEnabled = enabled)
    }

    fun setHorizontalNavigationEnabled(enabled: Boolean) {
        preferences.edit().putBoolean(KEY_HORIZONTAL_NAVIGATION, enabled).apply()
        _settings.value = _settings.value.copy(showHorizontalNavigation = enabled)
    }

    fun setShowSymbolsInSentenceBar(enabled: Boolean) {
        preferences.edit().putBoolean(KEY_SHOW_SYMBOLS_IN_SENTENCE, enabled).apply()
        _settings.value = _settings.value.copy(showSymbolsInSentenceBar = enabled)
    }

    fun setItemsToGenerate(count: Int) {
        val validCount = count.coerceIn(1, 50)
        preferences.edit().putInt(KEY_ITEMS_TO_GENERATE, validCount).apply()
        _settings.value = _settings.value.copy(itemsToGenerate = validCount)
    }

    fun setMaxBoardCapacity(count: Int) {
        val validCount = count.coerceIn(50, 1000)
        preferences.edit().putInt(KEY_MAX_BOARD_CAPACITY, validCount).apply()
        _settings.value = _settings.value.copy(maxBoardCapacity = validCount)
    }

    fun setDisabilityType(type: com.example.myaac.data.model.DisabilityType) {
        preferences.edit().putString(KEY_DISABILITY_TYPE, type.name).apply()
        _settings.value = _settings.value.copy(disabilityType = type)
        applyPersonalization(type)
    }

    fun setSymbolLibrary(library: String) {
        preferences.edit().putString(KEY_SYMBOL_LIBRARY, library).apply()
        _settings.value = _settings.value.copy(symbolLibrary = library)
    }

    fun setShowHomeOnStartup(show: Boolean) {
        preferences.edit().putBoolean(KEY_SHOW_HOME_ON_STARTUP, show).apply()
        _settings.value = _settings.value.copy(showHomeOnStartup = show)
    }

    fun setHomeButtons(buttonsJson: String) {
        preferences.edit().putString(KEY_HOME_BUTTONS, buttonsJson).apply()
        _settings.value = _settings.value.copy(homeButtonsJson = buttonsJson)
    }

    fun getAppShortcuts(): List<AppShortcut> {
        val json = _settings.value.appShortcutsJson
        if (json.isEmpty()) return emptyList()
        return try {
            val type = object : TypeToken<List<AppShortcut>>() {}.type
            gson.fromJson(json, type)
        } catch (e: Exception) {
            emptyList()
        }
    }

    fun setAppShortcuts(shortcuts: List<AppShortcut>) {
        val json = gson.toJson(shortcuts)
        preferences.edit().putString(KEY_APP_SHORTCUTS, json).apply()
        _settings.value = _settings.value.copy(appShortcutsJson = json)
    }

    // Prediction settings
    fun setPredictionEnabled(enabled: Boolean) {
        preferences.edit().putBoolean(KEY_PREDICTION_ENABLED, enabled).apply()
        _settings.value = _settings.value.copy(predictionEnabled = enabled)
    }

    fun setAiPredictionEnabled(enabled: Boolean) {
        preferences.edit().putBoolean(KEY_AI_PREDICTION_ENABLED, enabled).apply()
        _settings.value = _settings.value.copy(aiPredictionEnabled = enabled)
    }

    fun setPredictionCount(count: Int) {
        val validCount = count.coerceIn(3, 8)
        preferences.edit().putInt(KEY_PREDICTION_COUNT, validCount).apply()
        _settings.value = _settings.value.copy(predictionCount = validCount)
    }

    fun setShowSymbolsInPredictions(show: Boolean) {
        preferences.edit().putBoolean(KEY_SHOW_SYMBOLS_IN_PREDICTIONS, show).apply()
        _settings.value = _settings.value.copy(showSymbolsInPredictions = show)
    }

    fun setLearnFromUsage(learn: Boolean) {
        preferences.edit().putBoolean(KEY_LEARN_FROM_USAGE, learn).apply()
        _settings.value = _settings.value.copy(learnFromUsage = learn)
    }

    fun setAutoGrammarCheck(enabled: Boolean) {
        preferences.edit().putBoolean(KEY_AUTO_GRAMMAR_CHECK, enabled).apply()
        _settings.value = _settings.value.copy(autoGrammarCheck = enabled)
    }

    private fun applyPersonalization(type: com.example.myaac.data.model.DisabilityType) {
        // Simple personalization logic
        when (type) {
            com.example.myaac.data.model.DisabilityType.VISUAL_IMPAIRMENT -> {
                if (_settings.value.textScale < 1.3f) setTextScale(1.3f)
            }
            com.example.myaac.data.model.DisabilityType.MOTOR_IMPAIRMENT -> {
                // Could increase button sizes in the future, for now maybe ensure text isn't too small
                if (_settings.value.textScale < 1.1f) setTextScale(1.1f)
            }
            else -> {}
        }
    }
    
    companion object {
        private const val KEY_LANGUAGE = "language_code"
        private const val KEY_TEXT_SCALE = "text_scale"
        private const val KEY_TTS_RATE = "tts_rate"
        private const val KEY_DISABILITY_TYPE = "disability_type"
        private const val KEY_LOCATION_SUGGESTIONS = "location_suggestions_enabled"
        private const val KEY_HORIZONTAL_NAVIGATION = "horizontal_navigation_enabled"
        private const val KEY_SHOW_SYMBOLS_IN_SENTENCE = "show_symbols_in_sentence"
        private const val KEY_ITEMS_TO_GENERATE = "items_to_generate"
        private const val KEY_MAX_BOARD_CAPACITY = "max_board_capacity"
        private const val KEY_FONT_FAMILY = "font_family"
        private const val KEY_DISPLAY_SCALE = "display_scale"
        private const val KEY_USE_SYSTEM_SETTINGS = "use_system_settings"
        private const val KEY_SYMBOL_LIBRARY = "symbol_library"
        private const val KEY_SHOW_HOME_ON_STARTUP = "show_home_on_startup"
        private const val KEY_HOME_BUTTONS = "home_buttons_json"
        private const val KEY_APP_SHORTCUTS = "app_shortcuts_json"
        // Prediction keys
        private const val KEY_PREDICTION_ENABLED = "prediction_enabled"
        private const val KEY_AI_PREDICTION_ENABLED = "ai_prediction_enabled"
        private const val KEY_PREDICTION_COUNT = "prediction_count"
        private const val KEY_SHOW_SYMBOLS_IN_PREDICTIONS = "show_symbols_in_predictions"
        private const val KEY_LEARN_FROM_USAGE = "learn_from_usage"
        private const val KEY_AUTO_GRAMMAR_CHECK = "auto_grammar_check"
    }
}

File: web/src/data/settingsRepository.ts
import type { AppShortcut } from '../models';
import { DisabilityType } from '../models';
import { db } from './db';
import { DEFAULT_APP_SETTINGS, SETTINGS_ROW_ID } from './settingsTypes';
import type { AppSettings, SettingsRow } from './settingsTypes';

// Mirrors app/src/main/java/com/example/myaac/data/repository/SettingsRepository.kt.
//
// The Kotlin repository keeps a `SharedPreferences` instance plus an in-memory
// `MutableStateFlow<AppSettings>` it updates on every setter call. This module is the
// storage-and-persistence half of that: every function here reads/writes the single
// `settings` row in Dexie (see db.ts) and returns the resulting `AppSettings` snapshot.
// The reactive half (the StateFlow equivalent that UI code subscribes to) is
// `store/settingsStore.ts`, which calls into these functions and republishes the
// result — deliberately split the same way Kotlin splits "SharedPreferences I/O" from
// "the StateFlow observers actually read".
//
// Every one of SettingsRepository's ~24 SharedPreferences keys/setters is ported
// 1:1 below; see this module's export list. One intentional omission: the Kotlin
// `loadSettings()` has a one-time backward-compatibility shim that reads a legacy
// `max_board_items` key (from a preferences schema that predates the
// itemsToGenerate/maxBoardCapacity split) to compute defaults. That shim exists only
// to carry old *SharedPreferences* installs forward; per the migration plan there is
// no Room/SharedPreferences -> Dexie data migration; the hybrid app ships as a fresh
// install, so that legacy key can never be present here and the shim is not ported.

function coerceIn(value: number, min: number, max: number): number {
  return Math.min(Math.max(value, min), max);
}

/** Mirrors `loadSettings()` for the initial read, creating the row with defaults on first run. */
export async function loadSettings(): Promise<AppSettings> {
  const row = await db.settings.get(SETTINGS_ROW_ID);
  if (row) {
    const { id: _id, ...settings } = row;
    return settings;
  }
  await db.settings.put({ id: SETTINGS_ROW_ID, ...DEFAULT_APP_SETTINGS });
  return { ...DEFAULT_APP_SETTINGS };
}

async function updateSettings(patch: Partial<AppSettings>): Promise<AppSettings> {
  const current = await loadSettings();
  const updated: AppSettings = { ...current, ...patch };
  const row: SettingsRow = { id: SETTINGS_ROW_ID, ...updated };
  await db.settings.put(row);
  return updated;
}

export async function setLanguage(code: string): Promise<AppSettings> {
  return updateSettings({ languageCode: code });
}

export async function setTextScale(scale: number): Promise<AppSettings> {
  return updateSettings({ textScale: scale });
}

export async function setFontFamily(family: string): Promise<AppSettings> {
  return updateSettings({ fontFamily: family });
}

export async function setDisplayScale(scale: number): Promise<AppSettings> {
  return updateSettings({ displayScale: scale });
}

export async function setUseSystemSettings(useSystem: boolean): Promise<AppSettings> {
  return updateSettings({ useSystemSettings: useSystem });
}

export async function setTtsRate(rate: number): Promise<AppSettings> {
  return updateSettings({ ttsRate: rate });
}

export async function setLocationSuggestionsEnabled(enabled: boolean): Promise<AppSettings> {
  return updateSettings({ locationSuggestionsEnabled: enabled });
}

export async function setHorizontalNavigationEnabled(enabled: boolean): Promise<AppSettings> {
  return updateSettings({ showHorizontalNavigation: enabled });
}

export async function setShowSymbolsInSentenceBar(enabled: boolean): Promise<AppSettings> {
  return updateSettings({ showSymbolsInSentenceBar: enabled });
}

export async function setLandscapeBigSentence(enabled: boolean): Promise<AppSettings> {
  return updateSettings({ landscapeBigSentence: enabled });
}

/** Clamped to [1, 50], mirroring `setItemsToGenerate`. */
export async function setItemsToGenerate(count: number): Promise<AppSettings> {
  return updateSettings({ itemsToGenerate: coerceIn(count, 1, 50) });
}

/** Clamped to [50, 1000], mirroring `setMaxBoardCapacity`. */
export async function setMaxBoardCapacity(count: number): Promise<AppSettings> {
  return updateSettings({ maxBoardCapacity: coerceIn(count, 50, 1000) });
}

/** Minimum textScale each DisabilityType enforces, mirroring `applyPersonalization`. */
function minTextScaleFor(type: DisabilityType): number | null {
  switch (type) {
    case DisabilityType.VISUAL_IMPAIRMENT:
      return 1.3;
    case DisabilityType.MOTOR_IMPAIRMENT:
      return 1.1;
    default:
      return null;
  }
}

/** Mirrors `setDisabilityType(type)`, including its `applyPersonalization` side effect. */
export async function setDisabilityType(type: DisabilityType): Promise<AppSettings> {
  const updated = await updateSettings({ disabilityType: type });
  const minTextScale = minTextScaleFor(type);
  if (minTextScale !== null && updated.textScale < minTextScale) {
    return setTextScale(minTextScale);
  }
  return updated;
}

export async function setSymbolLibrary(library: string): Promise<AppSettings> {
  return updateSettings({ symbolLibrary: library });
}

export async function setShowHomeOnStartup(show: boolean): Promise<AppSettings> {
  return updateSettings({ showHomeOnStartup: show });
}

export async function setHomeButtons(buttonsJson: string): Promise<AppSettings> {
  return updateSettings({ homeButtonsJson: buttonsJson });
}

/** Mirrors `getAppShortcuts(): List<AppShortcut>` (parses the current settings row's JSON blob). */
export async function getAppShortcuts(): Promise<AppShortcut[]> {
  const settings = await loadSettings();
  return parseAppShortcuts(settings.appShortcutsJson);
}

/** Pure JSON-parsing half of `getAppShortcuts`, exposed so the Zustand store can reuse it without a DB round-trip. */
export function parseAppShortcuts(json: string): AppShortcut[] {
  if (json.length === 0) return [];
  try {
    const parsed: unknown = JSON.parse(json);
    return Array.isArray(parsed) ? (parsed as AppShortcut[]) : [];
  } catch {
    return [];
  }
}

/** Mirrors `setAppShortcuts(shortcuts)`. */
export async function setAppShortcuts(shortcuts: AppShortcut[]): Promise<AppSettings> {
  return updateSettings({ appShortcutsJson: JSON.stringify(shortcuts) });
}

// Prediction settings

export async function setPredictionEnabled(enabled: boolean): Promise<AppSettings> {
  return updateSettings({ predictionEnabled: enabled });
}

export async function setAiPredictionEnabled(enabled: boolean): Promise<AppSettings> {
  return updateSettings({ aiPredictionEnabled: enabled });
}

/** Clamped to [3, 8], mirroring `setPredictionCount`. */
export async function setPredictionCount(count: number): Promise<AppSettings> {
  return updateSettings({ predictionCount: coerceIn(count, 3, 8) });
}

export async function setShowSymbolsInPredictions(show: boolean): Promise<AppSettings> {
  return updateSettings({ showSymbolsInPredictions: show });
}

export async function setLearnFromUsage(learn: boolean): Promise<AppSettings> {
  return updateSettings({ learnFromUsage: learn });
}

export async function setAutoGrammarCheck(enabled: boolean): Promise<AppSettings> {
  return updateSettings({ autoGrammarCheck: enabled });
}
